File: legacy/legacy-mahabharata-ios/Mahabharata/Model/DataClasses/Visual/Animations/ScaleAnim.java
package com.fulldome.mahabharata.model.visual.animation;

import com.fulldome.mahabharata.model.visual.Layer;

public class ScaleAnim extends LayerAnim {
	private float scaleX;
	private float scaleY;
	private float pivotX;
	private float pivotY;

	public ScaleAnim() {
	}

	public ScaleAnim(float scaleX, float scaleY, float pivotX, float pivotY) {
		this.scaleX = scaleX;
		this.scaleY = scaleY;
		this.pivotX = pivotX;
		this.pivotY = pivotY;
	}

	public float getScaleX() {
		return scaleX;
	}

	public float getScaleY() {
		return scaleY;
	}

	public float getPivotX() {
		return pivotX;
	}

	public float getPivotY() {
		return pivotY;
	}

	@Override
	protected ScaleAnim interpolate(LayerAnim endAnim, float fraction) {
		ScaleAnim end = (ScaleAnim) endAnim;
		return new ScaleAnim(Layer.FLOAT_EVALUATOR.evaluate(fraction, getScaleX(), end.getScaleX()),
				Layer.FLOAT_EVALUATOR.evaluate(fraction, getScaleY(), end.getScaleY()),
				Layer.FLOAT_EVALUATOR.evaluate(fraction, getPivotX(), end.getPivotX()),
				Layer.FLOAT_EVALUATOR.evaluate(fraction, getPivotY(), end.getPivotY()));
	}

	@Override
	public void apply(Layer.ViewData data, int width, int height) {
		data.getMatrix().postTranslate(-width * pivotX, -height * pivotY);
		data.getMatrix().postScale(scaleX, scaleY);
		data.getMatrix().postTranslate(width * pivotX, height * pivotY);
	}
}

File: legacy/legacy-bhagavadgita.comics-ios-swift/Mahabharata/Model/DataClasses/Visual/Animations/ScaleAnim.swift
//
//  Comics.swift
//  Mahabharata
//
//  Created by Roman Developer on 11/28/17.
//  Copyright © 2017 Iron Water Studio. All rights reserved.
//

import UIKit

class ScaleAnim: Anim {
	let scaleX: Double
	let scaleY: Double
	let pivotX: Double
	let pivotY: Double
	
	private enum CodingKeys: String, CodingKey {
		case scaleX
		case scaleY
		case pivotX
		case pivotY
	}
	
	required init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		self.scaleX = (try? container.decode(Double.self, forKey: .scaleX)) ?? 0
		self.scaleY = (try? container.decode(Double.self, forKey: .scaleY)) ?? 0
		self.pivotX = (try? container.decode(Double.self, forKey: .pivotX)) ?? 0
		self.pivotY = (try? container.decode(Double.self, forKey: .pivotY)) ?? 0

		try super.init(from: decoder)
	}
	
	init(scaleX: Double, scaleY: Double, pivotX: Double, pivotY: Double) {
		self.scaleX = scaleX
		self.scaleY = scaleY
		self.pivotX = pivotX
		self.pivotY = pivotY
		
		super.init()
	}
	
	override var description: String {
		return [
			super.description,
			"ScaleX: \(scaleX)",
			"ScaleY: \(scaleY)",
			"PivotX: \(pivotX)",
			"PivotY: \(pivotY)"
			].joined(separator: ", ")
	}
	
	//MARK: - Overrides
	override func interpolate(endAnim: Anim, fraction: Double) -> Anim {
		let end = endAnim as! ScaleAnim
		return ScaleAnim(scaleX: LinearInterpolation.evaluate(fraction: fraction, startValue: self.scaleX, endValue: end.scaleX),
						 scaleY: LinearInterpolation.evaluate(fraction: fraction, startValue: self.scaleY, endValue: end.scaleY),
		                 pivotX: LinearInterpolation.evaluate(fraction: fraction, startValue: self.pivotX, endValue: end.pivotX),
		                 pivotY: LinearInterpolation.evaluate(fraction: fraction, startValue: self.pivotY, endValue: end.pivotY))
	}
	
	override func apply(to data: Layer.ViewData, width: Int, height: Int) {
		//Because we have system anchor point == (0.5, 0.5)
		let pivotX = self.pivotX - 0.5
		let pivotY = self.pivotY - 0.5
		
		data.matrix = CATransform3DTranslate(data.matrix, CGFloat(Double(width) * pivotX), CGFloat(Double(height) * pivotY), 0)
		data.matrix = CATransform3DScale(data.matrix, CGFloat(self.scaleX), CGFloat(self.scaleY), 1)
		data.matrix = CATransform3DTranslate(data.matrix, CGFloat(-Double(width) * pivotX), CGFloat(-Double(height) * pivotY), 0)
	}
}

